src/Microsoft.Data/DataFrame.cs

```diff
@@ -33,6 +36,273 @@ public DataFrame(IList<BaseColumn> columns)
+                        copied = byteValueBuffer.TryCopyTo(dataFrameMemory) && byteNullBitMapBuffer.TryCopyTo(nullBitMapMemory);
+                        if (copied)
+                        {
+                            dataFrameColumn = new PrimitiveColumn<byte>(field.Name, dataFrameMemory, nullBitMapMemory, arrowArray.Length);
+                        }
+                        break;
+                    case ArrowTypeId.Date32:
+                    case ArrowTypeId.Date64:
+                    case ArrowTypeId.Dictionary:
+                    case ArrowTypeId.FixedSizedBinary:
+                    case ArrowTypeId.HalfFloat:
+                    case ArrowTypeId.Interval:
+                    case ArrowTypeId.List:
+                    case ArrowTypeId.Map:
+                    case ArrowTypeId.Null:
+                    case ArrowTypeId.Struct:
+                    case ArrowTypeId.Time32:
+                    case ArrowTypeId.Time64:
+                    case ArrowTypeId.Timestamp:
+                    case ArrowTypeId.Union:
+                    default:
+                        throw new NotImplementedException(nameof(fieldType.Name));
+                }
+                _table.InsertColumn(ColumnCount, dataFrameColumn);
+                fieldIndex++;
+            }
+        }
+
+        public IEnumerable<RecordBatch> AsArrowRecordBatch()
+        {
+            Apache.Arrow.Schema.Builder schemaBuilder = new Apache.Arrow.Schema.Builder();
+            Field.Builder fieldBuilder = new Field.Builder();
+            // TODO: Sanity check. Some column types are not Arrow Compatible
+            // Derived column types will NOT be supported
+
+            // Find the number of DataFrameBuffers in each column. All columns have the same number of buffers
+            int numberOfColumns = ColumnCount;
+            // Assuming that all columns are Arrow compatible 
+            int numberOfRecordBatches = 0;
+            for (int i = 0; i < numberOfColumns; i++)
+            {
+                BaseColumn column = Column(i);
+                switch (column)
+                {
+                    case PrimitiveColumn<bool> boolColumn:
+                        numberOfRecordBatches = boolColumn.NumberOfBuffers;
+                        Field boolField = fieldBuilder.Name(boolColumn.Name).Nullable(true).DataType(BooleanType.Default).Build();
+                        schemaBuilder.Field(boolField);
+                        break;
+                    case PrimitiveColumn<byte> byteColumn:
+                    case PrimitiveColumn<char> charColumn:
+                    case PrimitiveColumn<decimal> decimalColumn:
+                    case PrimitiveColumn<sbyte> sbyteColumn:
+                        throw new NotImplementedException(nameof(byteColumn.DataType));
+                    case PrimitiveColumn<double> doubleColumn:
+                        numberOfRecordBatches = doubleColumn.NumberOfBuffers;
```

Comment:
The fact that you need to cast to a `PrimitiveColumn<double>` in order to access `NumberOfBuffers` kind of smells here.

Maybe `BaseColumn` should have a `ToArrow` method?

Reply:
Yup yup, I agree. I realized that too late and since this was just a hack patch, I just put it up.